### DISCUSSION TITLE
Color Picker for Color Blending FX

### DISCUSSION THREAD

**Discussion by KaseyFarron** [opentoonz/opentoonz#6828] [Ideas]
**Describe the solution you'd like**
I think it would be a little faster to use the Color Blending FX if we could color pick directly on the canvas instead of inputting the index numbers manually.

![image](https://user-images.githubusercontent.com/55353161/104315487-efebb780-54d2-11eb-887a-91705b2f8ac5.png)

**Comment by RodneyBaker** [comment]
I'm trying to imagine what the logic behind the code would have to do in order to implement this.
Here's something of a guess:

1) Using the current palette (obtained through node's input port reference)...
2) Sort the palette by color assignment and index number
3) Using color picker of gizmo of Color Blending FX...
4) Select a single color (or more complexly a range of colors from an area
5) Compare list of color picked color values to that of the palette color values
6) Return all color style index numbers that match
7) Populate the Color Blending::Color Indexes field with all the numbers discovered
8) Store these values for the field and prepare for the next user color pick
9) Append and then sort the values of subsequent color picks
10) In collecting numbers, duplicate colors must be allowed in order to be able to present indexes that contain the same color values.
11) A methodology for dealing with colors picked not currently included in the inputed palette/palettes would need to be devised;  for example if the color picked isn't in the palette then create a new color swatch containing that color and add its index into the Color Blending Color Index field.  This especially true as colors presented on the canvas may originate in different palettes and even in some cases not in any currently existing palette.

The hardest part to implement in this is likely the introduction of the color picker tool for FX which at this time does not yet exist.  Some code from the standard color pickers might likely be used to gain this or similar functionality.

Some logical extensions of this process present themselves such as the ability to target or exclude specific palettes for use in the blending which currently is provided by reference via the input port of the Color Blending node. 

A logical next step (feature request) toward implementation of this feature might be to implement a generalized color picker tool/widget for use with FX.


Added:  With further reflection...
It seems to me that there may be a more direct way to implement a copy/paste mechanism to use the current Color Picker to paste in values.  We might not have the color pick icon (and new color pick functionality) but it would allow for use of current color picking tools.

We almost... ALMOST... have this capability via Passive Pick.
![image](https://user-images.githubusercontent.com/2576432/104380499-b32eb900-54f0-11eb-924d-09192aa2941b.png)

Passive pick of course does not seek out current color styles that match the picked color.  It does however adjust the current color swatch to the passively picked color which could in turn be somehow transferred into the field for color blending...

**Comment by DarrenTAnims** [comment]
This should be possible and not too tricky to implement, I'd think. You've made it sound really complicated Rodney, but there's already a palette picker tool, so we'd just re-use that code, I'm sure.
But I like this idea. Adding a palette picker would be useful for a few of the effects and save me having to find the index number by hand, so would save time and be more accurate. I often type in the wrong number.
